scripts/migrate.py

```diff
@@ -0,0 +1,223 @@
+    logger.info("Initializing")
+
+    args = get_args()
+
+    if args.dryrun:
+        logger.info(" ---- DRYRUN MODE - No changes will be written to output file {}. --\n".format(args.outfile))
+
+    logger.info('Loading state file: ' + args.statefile)
+    if (args.statefile.startswith('s3://')):
+        tfstate = read_state_from_s3(args, logger)
+    else:
+        tfstate = read_state_from_file(args, logger)
+
+    if not args.dryrun and (args.statefile == args.outfile):
+        basename = os.path.basename(args.statefile)
+        backupfile = "{}.premigrate.{}".format(basename, datetime.now().strftime("%Y%m%d-%H%M%S"))
+        logger.info("Backing up {} to {}".format(args.statefile, backupfile))
+        write_state_to_file(backupfile, logger, tfstate)
+
+    if version.parse(tfstate['terraform_version']) < version.parse('0.12.1'):
+        logger.error("This migration script only works on state files created by Terraform 0.12.1+")
+        exit(1)
+
+    made_changes = False
+    # Change older version number to one that supports for_each
+    if version.parse(tfstate['terraform_version']) < version.parse('0.12.21'):
+        logger.info("Changing terraform_version from {} to {}.".format(tfstate['terraform_version'], '0.12.21'))
+        if not args.dryrun:
+            tfstate['terraform_version'] = '0.12.21'
+            made_changes = True
+
+    schema_index_map = get_schema_map(logger, tfstate)
+
+    made_changes |= change_resource_indices(args, logger, tfstate, 'template_file', 'bucket_policy', schema_index_map)
+    made_changes |= change_resource_indices(args, logger, tfstate, 'aws_s3_bucket', 'apiary_data_bucket', schema_index_map)
+    made_changes |= change_resource_indices(args, logger, tfstate, 'aws_s3_bucket_inventory', 'apiary_bucket', schema_index_map)
+    made_changes |= change_resource_indices(args, logger, tfstate, 'aws_s3_bucket_public_access_block', 'apiary_bucket', schema_index_map)
+    made_changes |= change_resource_indices(args, logger, tfstate, 'aws_s3_bucket_notification', 'data_events', schema_index_map)
+    made_changes |= change_resource_indices(args, logger, tfstate, 'aws_s3_bucket_metric', 'paid_metrics', schema_index_map)
+    made_changes |= change_resource_indices(args, logger, tfstate, 'aws_sns_topic', 'apiary_data_events', schema_index_map)
+
+    if not made_changes:
+        logger.info('State file was already migrated, no changes made, not saving to {}'.format(args.outfile))
+    else:
+        logger.info("Saving migrated state to {}".format(args.outfile))
+        if not args.dryrun:
+            if (args.outfile.startswith('s3://')):
+                write_state_to_s3(args, logger, tfstate)
+            else:
+                write_state_to_file(args.outfile, logger, tfstate)
+
+    if args.dryrun:
+        logger.info(" ---- DRYRUN MODE - No changes were written to output file {}. --".format(args.outfile))
+
+
+if __name__ == '__main__':
+    main()
```

Comment:
Add a [closing line break](https://stackoverflow.com/questions/5813311/no-newline-at-end-of-file) to EOF please.

Reply:
done